OpenMTRDemo/Forms/ExpandedImageForm.cs

```diff
@@ -1,23 +1,148 @@
+        {
+            if (_loadSaveDialog.openBrowser.ShowDialog() == DialogResult.OK)
+            {
+                Source = new Mat(_loadSaveDialog.openBrowser.FileName);
+                Render();
+                SetDisableableControls(true);
+            }
+        }
+
+        private void Render()
+        {
+            Image = Source.Clone();
+            ApplyFilters(Image);
+            OutputImageBox.Image = DemoUtilities.MatToBitmap(Image);
+        }
+
+        private void ApplyFilters(Mat imageToFilter)
+        {
+            foreach (string filter in FilterListBox.SelectedItems)
+            {
+                switch (filter)
+                {
+                    case "Black and White":
+                        Cv2.CvtColor(imageToFilter, imageToFilter, ColorConversionCodes.BGR2GRAY);
+                        break;
+                    case "Gaussian Blur":
+                        Cv2.GaussianBlur(imageToFilter, imageToFilter, new Size(3, 3), 0, 0, BorderTypes.Default);
+                        break;
+                    case "Edge Finding":
+                        if (cannyRadio.Checked)
+                        {
+                            Cv2.Canny(imageToFilter.Clone(), imageToFilter, CannyThreshold1Slider.Value, CannyThreshold2Slider.Value);
+                        }
+                        else
+                        {
+                            
+                            Cv2.Sobel(imageToFilter, imageToFilter, MatType.CV_8U, xorder: 1, yorder: 0, ksize: -1);
+                        }
+                        break;
+                }
+            }
+        }
+
+        private void filterListBox_SelectedIndexChanged(object sender, EventArgs e)
+        {
+            edgeFindingBox.Enabled = FilterListBox.SelectedItems.Contains("Edge Finding");
+            Render();
+        }
+
+        public override void CloseToolStripMenuItem_Click(object sender, EventArgs e)
+        {
+            SetDisableableControls(false);
+            WidthTextBox.Text = "";
+            HeightTextBox.Text = "";
+            OutputImageBox.Image = null;
+        }
+
+        private void SetDisableableControls(bool state)
+        {
+            SaveToolStripMenuItem.Enabled = state;
+            CloseToolStripMenuItem.Enabled = state;
+            FilterListBox.Enabled = state;
+            edgeFindingBox.Enabled = state && FilterListBox.SelectedItems.Contains("Edge Finding");
+            cannySettingsPanel.Enabled = cannyRadio.Checked;
+        }
+
+        private void cannyThreshold_ValueChanged(object sender, EventArgs e)
+        {
+            int value;
+            try
+            {
+                value = (int)((NumericUpDown)sender).Value;
+                TrackBar reciever = (sender == CannyThreshold1Number) ? CannyThreshold1Slider : CannyThreshold2Slider;
+                reciever.Value = value;
+            }
+            catch (InvalidCastException)
```

Comment:
What is causing the invalid cast exception and is there a way to prevent the user from ever being able to cause this?